imap_processing/mag/l1a/mag_l1a_data.py

```diff
@@ -429,6 +477,467 @@ def to_signed16(n: int) -> int:
+            )
+
+        if len(vector_data) != width * 3 + 2 * has_range:
+            raise ValueError(
+                f"Invalid length {len(vector_data)} for vector data. Expected "
+                f"{width * 3} or {width * 3 + 2} if has_range."
+            )
+        padding = np.zeros(8 - (width % 8), dtype=np.uint8)
+
+        # take slices of the input data and pack from an array of bits to an array of
+        # uint8 bytes
+        x = np.packbits(np.concatenate((padding, vector_data[:width])))
+        y = np.packbits(np.concatenate((padding, vector_data[width : 2 * width])))
+        z = np.packbits(np.concatenate((padding, vector_data[2 * width : 3 * width])))
+
+        range_string = "".join([str(i) for i in vector_data[-2:]])
+
+        rng = int(range_string, 2) if has_range else 0
+
+        # Convert to signed integers using twos complement
+        signed_vals: np.ndarray = np.array(
+            [
+                MagL1a.twos_complement(x, width),
+                MagL1a.twos_complement(y, width),
+                MagL1a.twos_complement(z, width),
+                rng,
+            ],
+            dtype=np.int32,
+        )
+        return signed_vals
+
+    @staticmethod
+    def twos_complement(value: np.ndarray, bits: int) -> np.int32:
+        """
+        Compute the two's complement of an integer.
+
+        This function will return the two's complement of a given bytearray value.
+        The input value should be a bytearray or a numpy array of uint8 values.
+
+        If the integer with respect to the number of bits does not have a sign bit
+        set (first bit is 0), then the input value is returned without modification.
+
+        Parameters
+        ----------
+        value : numpy.ndarray
+            An array of bytes representing an integer. In numpy, this should be an
+            array of uint8 values.
+        bits : int
+            Number of bits to use for the 2's complement.
+
+        Returns
+        -------
+        numpy.int32
+            Two's complement of the input value, as a signed int.
+        """
+        integer_value = int.from_bytes(value, "big")
+        if (integer_value & (1 << (bits - 1))) != 0:
+            output_value = integer_value - (1 << bits)
+        else:
+            output_value = integer_value
+        return np.int32(output_value)
+
+    @staticmethod
+    def decode_fib_zig_zag(code: np.ndarray) -> int:
+        """
+        Decode a fibonacci and zig-zag encoded value.
+
+        Parameters
+        ----------
+        code : numpy.ndarray
+            The code to decode, in the form of an array of bits (eg [0, 1, 0, 1, 1]).
+            This should always end in 2 ones (which indicates the end of a fibonacci
+            encoding).
+
+        Returns
+        -------
+        value: int
+            Signed integer value, with fibonacci and zig-zag encoding removed.
+        """
```

Comment:
Could there be a case in which code is too short?
if len(code) < 2:
        raise ValueError(f"Input code too short: {code}")

Reply:
I think it is ok for the length to be equal to 2 (this is how you represent 0.) If it's shorter than two, then it will be caught with the existing check... although without the nice error message. I'll add that in, good suggestion